src/btf.cpp

```diff
@@ -303,8 +303,137 @@ std::string BTF::type_of(const btf_type *type, const std::string &field)
   return std::string("");
 }
 
-} // namespace bpftrace
+static bool btf_type_is_modifier(const struct btf_type *t)
+{
+  /* Some of them is not strictly a C modifier
+   * but they are grouped into the same bucket
+   * for BTF concern:
+   *   A type (t) that refers to another
+   *   type through t->type AND its size cannot
+   *   be determined without following the t->type.
+   *
+   * ptr does not fall into this bucket
+   * because its size is always sizeof(void *).
+   */
+  switch (BTF_INFO_KIND(t->info))
+  {
+  case BTF_KIND_TYPEDEF:
+  case BTF_KIND_VOLATILE:
+  case BTF_KIND_CONST:
+  case BTF_KIND_RESTRICT:
+    return true;
+  default:
+    return false;
+  }
+}
+
+const struct btf_type *BTF::btf_type_skip_modifiers(const struct btf_type *t)
+{
+  while (btf_type_is_modifier(t))
+  {
+    t = btf__type_by_id(btf, t->type);
+  }
+
+  return t;
+}
+
+SizedType BTF::get_stype(__u32 id)
+{
+  SizedType stype = SizedType(Type::none, 8);
+
+  const struct btf_type *t = btf__type_by_id(btf, id);
+
+  if (!t)
+    return stype;
+
+  t = btf_type_skip_modifiers(t);
+
+  stype.is_tracing = true;
+
+  if (btf_is_int(t) || btf_is_enum(t))
+  {
+    stype.type = Type::integer;
+  }
+  else if (btf_is_ptr(t))
+  {
+    stype.is_pointer = true;
+
+    /* get the pointer type.. */
+    t = btf__type_by_id(btf, t->type);
+    /* .. and skip the trash. */
+    t = btf_type_skip_modifiers(t);
+
+    if (btf_is_composite(t))
+    {
+      const char *cast = btf_str(btf, t->name_off);
+
+      if (cast)
+      {
+        std::string comp = btf_is_struct(t) ? "struct" : "union";
+
+        stype.type = Type::cast;
```

Comment:
Recently, Tyep::ctx is added (#1104). Type::ctx is almost the same as the Type::cast, but codegen emits a load instruction instead of bpf_probe_read() when accessing a value. In my understanding, the verifier allow for BPF program to access a field of argument directly in kfunc/kretfunc since it can ensure a type of argument thanks to the BTF. So using Type::ctx here (and in the FieldAnalyser) will eliminate one bpf_probe_read().

... but currently Type::ctx value cannot be stored in map (https://github.com/iovisor/bpftrace/commit/b7d2510b9a7a13f3e71685fa35fb52c50e265f4e). So we need to change a semantic analyzer in some ways.

Well, using Type::ctx can be made in another PR.

Reply:
right, I'm using Type::ctx in new version